drivers/can/CMakeLists.txt

```diff
@@ -14,6 +14,21 @@ zephyr_library_sources_ifdef(CONFIG_CAN_STM32FD      can_stm32fd.c)
 zephyr_library_sources_ifdef(CONFIG_CAN_STM32H7      can_stm32h7.c)
 zephyr_library_sources_ifdef(CONFIG_CAN_RCAR         can_rcar.c)
 
+if(CONFIG_CAN_NATIVE_LINUX)
+  if(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Linux)
+    zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/net/l2)
+    zephyr_library_compile_definitions(NO_POSIX_CHEATS)
+    zephyr_library_compile_definitions(_BSD_SOURCE)
+    zephyr_library_compile_definitions(_DEFAULT_SOURCE)
+    zephyr_library_sources(
+      can_native_linux.c
+      can_native_linux_adapt.c
+    )
+  else()
+    message(FATAL_ERROR "CONFIG_CAN_NATIVE_LINUX only available on Linux")
+  endif()
+endif()
```

Comment:
Not sure of the naming here. I think I may actually prefer `can_native_posix.c` - even though it is Linux-specific - and keep the cmake check in place. I'm afraid the `can_native_linux` naming will give us far too many support questions...

Reply:
I renamed it because of https://github.com/zephyrproject-rtos/zephyr/pull/37124#discussion_r676395876.

But I don't really care that much, I can rename it back to `native_posix` to avoid confusions. I realized that also the [Kconfig.board file](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_posix/Kconfig.board) only talks about Linux and not other POSIX-compatible systems.

Reply:
I think the driver should really be `can_socketcan`. SocketCAN has nothing to do with POSIX other than it's available on a POSIX compatible OS.

Reply:
@gmarull That would be even more confusing as it clashed with the Zephyr SocketCAN implementation: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/socketcan.h

Reply:
Our other drivers are named from the platform, they work on (e.g. `*_stm32,c` works on the STM32 platform). One could argue that the drivers, which work on the `native_posix` platform, should be named `*_native_posix`. This driver will never work on other platforms, but it is true it will not work on `native_posix` under all hosts.

Another possible name could be `can_native_posix_linux.c`?

Reply:
I agree that `can_socketcan` would be more confusing because our more or less compatible API in-tree. But as SocketCAN is *the* API for CAN in Linux, `can_native_linux` is maybe not too bad...

Reply:
So maybe `can_linux_socketcan`?

Reply:
Would be fine with me if @henrikbrixandersen agrees.

Reply:
I would prefer to keep `native_posix` as part of the name. We claim some level of Linux SocketCAN compatibility for the Zephyr SocketCAN implementation as well. Naming it `can_linux_socket.c` will be confusing.

I suggest `can_native_posix_linux.c` as a compromise between the suggestions so far.

Reply:
Ok, looks like there is [MacCAN](https://mac-can.github.io/). If there will ever be anyone who implements this, it could live in `can_native_posix_macos.c`. So I agree with the name for consistency reasons even though it is quite long.

Reply:
AFAIK, the driver uses the Linux SocketCAN interface internally to expose the Zephyr CAN API, so I'd expect the name to reflect that. `native_posix` implies it will run on native_posix, but it may not be true if we ever have e.g. `native_posix` on macOS. To clarify my argument, imagine we ever have a `native_windows` port, CAN drivers there won't be generic, we'll probably need `native_windows_peak`, `native_windows_kvaser`, etc.

Reply:
Exactly. This naming scheme matches our existing Zephyr driver naming schemes of `<"driver class">_<"platform">_<"backend">.c` (e.g. `counter_nrfx_rtc`, `counter_nrfx_timer`).

Reply:
Backend is `socketcan`, not Linux, on macOS it would be `maccan`.

Reply:
I've now named the files with the Linux SocketCAN glue code `can_native_posix_linux_socketcan.h/c` and the actual Zephyr driver implementation `can_native_posix_linux.c`. Hope that satisfies both of your requests.